kroxylicious-additional-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java

```diff
@@ -0,0 +1,130 @@
+import java.util.stream.Stream;
+
+import org.apache.kafka.common.record.Record;
+
+import io.kroxylicious.filter.encryption.KeyManager;
+import io.kroxylicious.filter.encryption.Receiver;
+import io.kroxylicious.filter.encryption.RecordEncryptionRequest;
+import io.kroxylicious.kms.service.Kms;
+import io.kroxylicious.kms.service.Serde;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * An implementation of {@link KeyManager} that uses envelope encryption, AES-GCM and stores the KEK id and encrypted DEK
+ * alongside the record ("in-band").
+ * @param <K> The type of KEK id.
+ * @param <E> The type of the encrypted DEK.
+ */
+public class InBandKeyManager<K, E> implements KeyManager<K> {
+
+    private final Kms<K, E> kms;
+    private final BufferPool bufferPool;
+    private final Serde<K> kekIdSerde;
+    private final Serde<E> edekSerde;
+
+    public InBandKeyManager(Kms<K, E> kms, BufferPool bufferPool) {
+        this.kms = kms;
+        this.bufferPool = bufferPool;
+        this.edekSerde = kms.edekSerde();
+        this.kekIdSerde = kms.keyIdSerde();
+    }
+
+    private CompletionStage<KeyContext<K>> currentDekContext(@NonNull K kekId) {
+        // TODO caching with expiry
+        // TODO count the number of encryptions
+        // TODO destroy DEK material
+        return kms.generateDekPair(kekId)
+                .thenApply(dekPair -> {
+                    E edek = dekPair.edek();
+                    short kekIdSize = (short) kekIdSerde.sizeOf(kekId);
+                    short edekSize = (short) edekSerde.sizeOf(edek);
+                    // TODO use buffer pool
+                    ByteBuffer prefix = ByteBuffer.allocate(
+                            Short.BYTES + // kekId size
+                                    kekIdSize + // the kekId
+                                    Short.BYTES + // DEK size
+                                    edekSize); // the DEK
+                    prefix.putShort(kekIdSize);
+                    kekIdSerde.serialize(kekId, prefix);
+                    prefix.putShort(edekSize);
+                    edekSerde.serialize(edek, prefix);
+                    prefix.flip();
+
+                    return new KeyContext<>(prefix,
+                            new AesGcmEncryptor(new AesGcmIvGenerator(new SecureRandom()), dekPair.dek()));
+                });
+    }
+
+    @NonNull
+    @Override
+    public CompletionStage<Void> encrypt(@NonNull K kekId,
+                                         @NonNull Stream<RecordEncryptionRequest> recordRequests,
+                                         @NonNull Receiver receiver) {
+        return currentDekContext(kekId).thenAccept(keyContext -> {
+            recordRequests.forEach(recordRequest -> {
+                // XXX accumulator
+                var output = bufferPool.acquire(keyContext.encodedSize(recordRequest.plaintextSize()));
+                try {
+                    // TODO encrypt null values
```

Comment:
I don't follow where you are going with this comment.  I expected that for Kafka's tombstone semantics, we need to propagate a null record to the broker as-is.  If we didn't do that we'd break compactions ability to delete the tombstones from the log.

https://kafka.apache.org/documentation/#compaction:~:text=A%20message%20with%20a%20key%20and%20a%20null%20payload

> Compaction also allows for deletes. A message with a key and a null payload will be treated as a delete from the log. Such a record is sometimes referred to as a tombstone. This delete marker will cause any prior message with that key to be removed (as would any new message with that key), but delete markers are special in that they will themselves be cleaned out of the log after a period of time to free up space. 

so, I'm think that records with null values should skip the encrypt stage altogether.    I'd do the short circuiting in the filter. The KeyManager won't know about them at all.

Or am I missing something?

Reply:
You're quite right, this _would_ break tombstone semantics.

I was thinking about the problem more from the PoV of what information we're exposing by _not_ encrypting null. If we're only encrypting non-null record values (not record keys) then it means that an adversary can all deletions in all topics in the clear.

Clearly, given that the broker has an interpretation of a null value in a compacted topic, there is a trade off between better confidentiality vs the ability to compact topics at all.

So for an MVP I think you're quite right that null should encrypt as null. 

The confidentiality could be improved in at least two ways:

1. Use deterministic encryption for record keys, so an adversary wouldn't see the plaintext of the key that we being deleted.

2. Use a deferred deletion mechanism, where we encrypt null values, but there exists some privileged application which is able to read the null value through the decryption filter and then produce an unencrypted null so that there's a real tombstone. This means that the time at which the broker and adversary observe a tombstone is some time after the time at which applications observe a tombstone, which is a defence against an adversary doing some metadata analysis. I.e. they can't reliably infer that Bob does action X and it results in a tombstone being emitted for record with key foo because of a large and random time delay between those two events.

Option 1 is a realistic future objective. Option 2 is more of a thought experiment.

Reply:
> So for an MVP I think you're quite right that null should encrypt as null.

Make sense.

> I was thinking about the problem more from the PoV of what information we're exposing by not encrypting null. If we're only encrypting non-null record values (not record keys) then it means that an adversary can all deletions in all topics in the clear.

You missed a word out from that sentence.  can **observe** all deletions?

> Use deterministic encryption for record keys, so an adversary wouldn't see the plaintext of the key that we being deleted.

I see that's technically possible but there would be challenges about how the record key encrypting key gets rotated.  I don't think it is possible without application downtime: rotate keys + reencrypt all existing keys.

> Use a deferred deletion mechanism, 

that's an interesting idea.  

> they can't reliably infer that Bob does action X and it results in a tombstone being emitted for record with key foo because of a large and random time delay between those two events.

Doesn't it just make things harder (rather than impossible) ? An adversary observing the log could still work Bob caused the tombstone.  They just need to correlate the real null for record R against the previous record for R (which must have been the encrypted null).